readme.md
# Super Metroid Zero Percent

A rom hack where you run the Super Metroid Game starting with full items/health/ammo and have to return everything on the planet zebes.

## Build

#### Setup tools

- Add files to ```tools\asar171\asar.exe``` from [Asar 1.71 at SMW Central](https://www.smwcentral.net/?p=section&a=details&id=19043) to patch assembly files
- Add files to ```tools\flips.exe``` from [Floating IPS (Flips) v1.31 at SMW Central](https://www.smwcentral.net/?p=section&a=details&id=11474) to patch IPS files

#### Rom Source

- Place SuperMetroid rom in ```rom.sfc``` alongside build.bat file

#### Compile

- Run ```build.bat``` and output will be in ```build/rom.sfc```

changelog.md
0.8.2

- made g4 door entrance a grey door
- started editing intro title screen
- allowed items to be dropped off visually
- prevented auto reserve when dropped off last reserve tank

- fixed bt spawn on collect item
- added right side ws e-tank when disabled 
- added lower norfair no spin jump activate pillar

0.8.1

- Changed filename name to Super Metroid Zero Percent
- Fixed bomb bug
- Fixed selection bug when removing an item you have selected
- Forced 0% ending
- made health not refill when collecting tanks

0.8.0

- Fixed gravity/varia

src/loadout.asm
arch snes.cpu
lorom

;; Code in bank 81 newgame
org $81B2CD
   JMP $EF1A

; Free space in 81:EF1A -> 81:FF00
; Free space in 81:FF60 -> 81:FFFF

org $81EF1A
new_heath:
   LDA #$05DB             ;\ 1499 = 05DB, 499 = 01F3, 199 = 00C7
   STA $09C4 ; [$7E:09C4]  ;} Health
   STA $09C2  ; [$7E:09C2]  ;/
   LDA #$0005             ;\  00E6   =  230  000A = 10
   STA $09C8 ; [$7E:09C8]  ;} Missiles
   STA $09C6 ; [$7E:09C6]  ;/
   LDA #$0005             ;\ = 0032 = 50
   STA $09CC ; [$7E:09CC]  ;} Super missiles
   STA $09CA ; [$7E:09CA]  ;/
   LDA #$0005             ;\ = 0032 = 50
   STA $09D0 ; [$7E:09D0]  ;} Power bombs
   STA $09CE ; [$7E:09CE]  ;/
   STZ $09D2 ; [$7E:09D2]  ; Currently selected HUD item
   LDA #$100F ; 1000000001111
   STA $09A8 ; set beams
   LDA #$100B ; 1000000001011 - not spazer
   STA $09A6 ; set beams
   LDA #$F32F ; collected items ; 1111001100101111
   STA $09A4 ; set items
   LDA #$F32F ; equited items
   STA $09A2 ; set items
   LDA #$0190
   STA $09D4 ; set reserve
   STA $09D6 ; set reserve

   LDA #$0001
   STA $09C0 ; set reserve mode to auto

   ; debug set minimum health/items/beams
;   LDA #$00C7
;   STA $09C4
;   STA $09C2
;   LDA #$0005
;   STA $09C8
;   STA $09C6
;   STA $09CC
;   STA $09CA
;   STA $09D0
;   STA $09CE
;   LDA #$1000
;   STA $09A8
;   STA $09A6
;   LDA #$0064
;   STA $09D4 ; set reserve
;   STA $09D6 ; set reserve

   ; toggle zebes awake
   LDA $7ED820    ;
   ORA #$0001   ;
   STA $7ED820    ;
   JMP $B30C ; back to new save file after ammo



src/new_game.asm
;;; VARIA new game hook: skips intro and customizes starting point
;;;
;;; compile with asar (https://www.smwcentral.net/?a=details&id=14560&p=section),
;;; or a variant of xkas that supports arch directive

arch snes.cpu
lorom

;;; CONSTANTS
!GameStartState = $7ED914


;;; HIJACKS (bank 82 init routines)

org $82801d
    jsl startup

org $828067
    jsl gameplay_start

;;; This skips the intro : game state 1F instead of 1E
org $82eeda
    db $1f

;;; DATA in bank A1 (start options)

org $a1f200
print "start_location: ", pc
start_location:
    ;; start location: $0000=Zebes Landing site, $fffe=Ceres,
    ;; otherwise hi byte is area and low is save index.
    ;; (use FFFE as Ceres special value because FFFF can be mistaken
    ;; for free space by solver/tracker)
    dw $0000			; defaults to landing site
opt_doors:
    ;; optional doors to open.
    ;; door ID is low byte PLM argument when editing doors in SMILE
    ;; terminate with $00
    db $10,$32			; defaults to red tower top+construction zone
    db $00

warnpc $a1f20f

;;; CODE in bank A1
org $a1f210
;;; zero flag set if we're starting a new game
;;; called from credits_varia as well
print "check_new_game: ", pc
check_new_game:
    ;; Make sure game mode is 1f
    lda $7e0998
    cmp #$001f : bne .end
    ;; check that Game time and frames is equal zero for new game
    ;; (Thanks Smiley and P.JBoy from metconst)
    lda $09DA
    ora $09DC
    ora $09DE
    ora $09E0
.end:
    rtl

startup:
    jsl check_new_game      : bne .end
    lda.l start_location    : beq .zebes
    cmp #$fffe              : beq .ceres
    ;; custom start point on Zebes
    pha
    and #$ff00 : xba : sta $079f ; hi byte is area
    pla
    and #$00ff : sta $078b      ; low byte is save index
    lda #$0000 : jsl $8081fa    ; wake zebes
.zebes:
    lda #$0005 : bra .store_state
.ceres:
    lda #$001f
.store_state:
    sta !GameStartState
.end:
    ;; run hijacked code and return
    lda !GameStartState
    rtl

gameplay_start:
    jsl check_new_game  : bne .end
    ;; Set doors to blue if necessary
    phx
    ldx #$0000
-
    lda.l opt_doors,x : and #$00ff
    beq .save			; end list
    phx
    jsl $80818e		    ; call bit index function, returns X=byte index, $05e7=bitmask
    ;; Set door in bitfield
    lda $7ED8B0,x : ora $05E7 : sta $7ED8B0,x
    plx
    inx : bra -		    ; next
.save:
    ;; Call the save code to create a new file
    plx
    lda $7e0952 : jsl $818000
.end:
    rtl

warnpc $a1f29f

;;; since this patch is always included, we can put utility
;;; routines for other patches here (in fixed locations)

!RNG		= $808111	; RNG function
!RNG_seed	= $05e5
!RTA_timer	= $05b8		; see tracking.asm

org $a1f2a0
;;; single use (will give the same result if called several times in the same frame)
;;; random function that leaves game rng untouched
;;; result in A
rand:
    phy
    lda !RNG_seed : pha             ; save current rand seed
    eor !RTA_timer : sta !RNG_seed  ; alter seed with frame counter
    jsl !RNG : tay                  ; call RNG and save result to Y
    pla : sta !RNG_seed             ; restore current rand seed
    tya                             ; get RNG result in A
    ply
    rtl

warnpc $a1f2ff

;;; patch morph+missile room state check
org $8fe652
morph_missile_check:
    ;; check that zebes is awake instead: works with both standard
    ;; start with wake_zebes.ips, and non standard start with wake
    ;; zebes forced from the start.
    lda #$0000 : jsl $808233
    bcc .not_awake
    bra .awake
org $8fe65f
.awake:
org $8fe666
.not_awake:

src/plm_spawn.asm
;;; Spawns additional PLMs into rooms without needing to repoint anything
;;;
;;; compile with asar (https://www.smwcentral.net/?a=details&id=14560&p=section),
;;; or a variant of xkas that supports arch directive

arch snes.cpu
lorom

;;; HIJACKS (room load routines in bank 82)
org $82e8d5
    jsl add_plms

org $82eb8b
    jsl add_plms

;;; DATA in bank 8F free space
org $8fe9a0
;;; additional zero-terminated PLM lists go here (written at ROM generation)
print "downwards plm_lists table start: ", pc
plm_lists:

;;; TEST
     dw $B76F
     db $0B, $09
     dw $0007
;;; 
;    dw  $0000                   ; PLM 1 ID
;    dw  $00,$00                 ; PLM 1 X, PLM 1 Y
;    dw  $0000                   ; PLM 1 argument
    ;; ...
;    dw  $0000                   ; PLM n ID
;    dw  $00,$00                 ; PLM n X, PLM n Y
;    dw  $0000                   ; PLM n argument
    
    dw  $0000                   ; PLM list 1 terminator
    ;; ...

plm_lists2:
    ;; Missile: EE52
    ;; Energy:  EED7

    ;; test
     dw $EED7 ;$EED7
     db $3C, $4A
     dw $0032

    ;; energy
     ;dw $EED7
     ;db $3C, $4A
     ;dw $0032
     
     ; morhping ball
     ;dw $EF23
     ;db $36, $4A
     ;dw $0004
     
     ; speed booster
;     dw $EEF7
;     db $36, $4A
;     dw $0004
     
     ; reserve booster
;     dw $EF7B
;     db $38, $48
;     dw $0004
     ; reserve booster
;     dw $EF7B
;     db $39, $48
;     dw $0004
     ; reserve booster
;     dw $EF7B
;     db $3A, $48
;     dw $0004
     ; reserve tank
;     dw $EF7B
;     db $3B, $48
;     dw $0004
     ; moprhing ball
;     dw $EF23
;     db $3B, $48
;    dw $0004
     
     dw $0000

;;; *** Non-overlap in this space has to be handled at ROM generation ***

org $8febf8

;;; Additional PLM definitions for rooms *going upwards* (written at ROM generation)
;;; 
;;; PLM lists are indexed by (Room ptr, Room State ptr, Entry Door ptr) and terminated by $0000
;;;
;;; Only Room is mandatory, if any other is $0000 it will be ignored.
;;; 
;;; List terminated by room ID $0000
room_plms:
    ;;  Room   State  Door   PLM list address
;;    dw  $0000, $0000, $0000, $0000
;;; TEST
;; room_plms:
     ;dw  $99bd, $99ca, $8b1a, plm_lists
     dw  $91f8, $0000, $0000, plm_lists2
;;; 
print "upwards room_plms table start: ", pc
;;; CODE in bank 8F
org $8ff300

add_plms:
    phx
    ;; iterate through room table
    ;; we use only X as index register because of no "LDA bbaaaa,Y" instruction
    ldx #room_plms
.room_loop:
    lda $8f0000,x   : beq .end
    cmp $079b       : beq .room_ok
    bra .next_entry
.room_ok:
    lda $8f0002,x   : beq .state_ok
    cmp $07bb       : beq .state_ok
    bra .next_entry
.state_ok:
    lda $8f0004,x   : beq .load
    cmp $078d       : beq .load
    bra .next_entry
.load:
    phx                         ; save our room table iterator
    lda $8f0006,x : tax         ; put PLM list address in X
.plm_loop:
    lda $8f0000,x   : beq .load_end
    jsl $84846a                 ; spawn room PLM. argument is X : PLM offset in bank 8F
    rep 6 : inx                 ; X += 6
    bra .plm_loop
.load_end:
    plx                         ; restore room table iterator
.next_entry:
    txa : sec : sbc #$0008 : tax ; X -= 8
    bra .room_loop
.end:
    plx
    ;; vanilla code (call door asm)
    jsl $8fe8a3
    rtl

warnpc $8ff3ff


src/max_ammo_display.asm
;;; Display in HUD the maximum number of ammo that samus can carry
;;;
;;; compile with asar (https://www.smwcentral.net/?a=details&id=14560&p=section),
;;; or a variant of xkas that supports arch directive

lorom
arch snes.cpu

;;; $9D98: Draw two HUD digits ;;;
;;; Parameters:
;;;     A: Number to draw                   ; 0 -> 99
;;;     X: HUD tilemap index                ; to index into $7E:C608
;;;     $00: Long pointer to digits tilemap ; $80:9DD3 (Artillery HUD digits tilemap)
!draw_two_HUD_digits   = $9D98

;;; $9D78: Draw three HUD digits ;;;
;;; Parameters:
;;;     A: Number to draw                   ; 0 -> 999
;;;     X: HUD tilemap index                ; to index into $7E:C608
;;;     $00: Long pointer to digits tilemap ; $80:9DD3 (Artillery HUD digits tilemap)
!draw_three_HUD_digits = $9D78

!samus_missiles           = $09C6
!samus_max_missiles       = $09C8
!samus_super_missiles     = $09CA
!samus_max_super_missiles = $09CC
!samus_power_bombs        = $09CE
!samus_max_power_bombs    = $09D0

;;;    $09D2: HUD item index
;;;    {
;;;        0: Nothing
;;;        1: Missiles
;;;        2: Super missiles
;;;        3: Power bombs
;;;        4: Grapple beam
;;;        5: X-ray
;;;    }
!hud_item_index            = $09D2
!hud_item_index_missile    = #$0001
!hud_item_index_super      = #$0002
!hud_item_index_power_bomb = #$0003

;;; see $7E:C608..C7: HUD tilemap. Not including top-most row (row 0)
!row1_missile_index      = #$0014
!row1_missile_tile0      = $7EC61C
!row1_missile_tile1      = $7EC61E
!row1_missile_tile2      = $7EC620
!row2_missile_tile0      = $7EC65C
!row2_missile_tile1      = $7EC65E
!row2_missile_tile2      = $7EC660
!row3_missile_index      = #$0094

!row1_super_index        = #$001A
!row1_super_tile0        = $7EC622
!row1_super_tile1        = $7EC624
!row1_super_tile2        = $7EC626
!row2_super_tile0        = $7EC664
!row2_super_tile1        = $7EC666
!row3_super_index_vanilla = #$009C
!row3_super_index_custom  = #$009A
!row3_super_tile0        = $7EC6A2

!row1_power_bomb_index   = #$0020
!row1_power_bomb_tile0   = $7EC628
!row1_power_bomb_tile1   = $7EC62A
!row1_power_bomb_tile2   = $7EC62C
!row2_power_bomb_tile0   = $7EC66A
!row2_power_bomb_tile1   = $7EC66C
!row3_power_bomb_index_vanilla = #$00A2
!row3_power_bomb_index_custom  = #$00A0
!row3_power_bomb_tile0   = $7EC6A8

;;; ;;; $9DBF: HUD digits tilemap ;;;
;;; {
;;; ; Artillery                0     1     2     3     4     5     6     7     8     9
;;; $80:9DD3             dw 2C09, 2C00, 2C01, 2C02, 2C03, 2C04, 2C05, 2C06, 2C07, 2C08
;;; }
!HUD_digits_tilemap_row3 = #$9DD3
!empty_tile = #$2C0F

;;; used to return values from functions which extract digits
!digit1 = $0012
!digit2 = $0014
!digit3 = $0016

;;; or mask applied for selected/unselected item in hud
!select_mask = $0018
!selected   = #$1000
!unselected = #$1400

;;; 8x8 mult => 16
!mult1 = $4202
!mult2 = $4203
!mult_result = $4216

;;; 16/8 div => 16,16
!dividend  = $4204
!divisor   = $4206
!quotient  = $4214
!remainder = $4216

;;; NOP
org $8099E1
	;; nop it:
        ;; $80:99E1 AD A3 99    LDA $99A3  [$80:99A3]  ;\
        ;; $80:99E4 8F 1C C6 7E STA $7EC61C[$7E:C61C]  ;|
        ;; $80:99E8 AD A5 99    LDA $99A5  [$80:99A5]  ;|
        ;; $80:99EB 8F 1E C6 7E STA $7EC61E[$7E:C61E]  ;} Write top row of missile icon
        ;; $80:99EF AD A7 99    LDA $99A7  [$80:99A7]  ;|
        ;; $80:99F2 8F 20 C6 7E STA $7EC620[$7E:C620]  ;/
        nop : nop : nop
        nop : nop : nop : nop
        nop : nop : nop
        nop : nop : nop : nop
        nop : nop : nop
        nop : nop : nop : nop

org $809ACE
        ;; nop it:
        ;; $80:9ACE 22 CF 99 80 JSL $8099CF            ; Add missiles to HUD tilemap
        nop : nop : nop : nop

org $809AD7
        ;; nop it:
        ;; $80:9AD7 22 0E 9A 80 JSL $809A0E            ; Add super missiles to HUD tilemap
        nop : nop : nop : nop

org $809AE0
        ;; nop it:
        ;; $80:9AE0 22 1E 9A 80 JSL $809A1E            ; Add power bombs to HUD tilemap
        nop : nop : nop : nop

;;; HIJACKS
org $809B1A
        ;; hijack vanilla super missile count drawing on row 3 to display on three digits when >= 100
        jsr super_missile_count_drawing_row3
	
org $809B28
        ;; hijack vanilla power bomb count drawing on row 3 to display on three digits when >= 100
        jsr power_bomb_count_drawing_row3

org $809C00
        ;; hijack into: $9B44: Handle HUD tilemap (HUD routine when game is paused/running) ;;;
        ;; ; Handle Samus' missiles
        ;; $80:9BFB A9 D3 9D    LDA #$9DD3             ;\
        ;; $80:9BFE 85 00       STA $00    [$7E:0000]  ;} $00 = pointer to digit tiles
        ;; $80:9C00 AD C8 09    LDA $09C8  [$7E:09C8]  ;\
        ;; $80:9C03 F0 11       BEQ $11    [$9C16]     ;} If [Samus' max missiles] != 0:
        jmp hud_ammo_drawing


;;; parameters set by vanilla before all hijacks:
;;; LDA !HUD_digits_tilemap_row3
;;; STA $00    [$7E:0000]
org $80CDA0

;;; parameters set before hijack:
;;; LDX !row3_super_index_vanilla
;;; LDA !samus_super_missiles
super_missile_count_drawing_row3:
	cmp #$0064
	bpl .three_digits
.two_digits
        jsr !draw_two_HUD_digits
	;; display empty tile in front of two digits
        lda !empty_tile
        sta !row3_super_tile0
	bra .end
.three_digits
	ldx !row3_super_index_custom
        jsr !draw_three_HUD_digits
.end
        rts

;;; parameters set before hijack:
;;; LDX !row3_power_bomb_index_vanilla
;;; LDA !samus_power_bombs
power_bomb_count_drawing_row3:
	cmp #$0064
	bpl .three_digits
.two_digits
        jsr !draw_two_HUD_digits
	;; display empty tile in front of two digits
        lda !empty_tile
        sta !row3_power_bomb_tile0
	bra .end
.three_digits
	ldx !row3_power_bomb_index_custom
        jsr !draw_three_HUD_digits
.end
        rts
 
;;; hijack into: $9B44: Handle HUD tilemap (HUD routine when game is paused/running) ;;;
;;; draw all three rows for missile/super/powerbomb as we skip that part in vanilla code
hud_ammo_drawing:
        pha 
        phx 
        phy 
        lda !samus_max_missiles
        beq .no_missile
        jsr draw_missile
        bra .supers
.no_missile:
	jsr clear_missile
.supers:
        lda !samus_max_super_missiles
        beq .no_super
        jsr draw_super
        bra .power_bombs
.no_super:
        jsr clear_super
.power_bombs:
        lda !samus_max_power_bombs
        beq .no_power_bomb
        jsr draw_power_bomb
        bra .finish_draw
.no_power_bomb:
        jsr clear_power_bomb
.finish_draw:
        ply 
        plx 
        pla 
        jmp $9C55

draw_missile:
        jsr extract_three_digits
        lda !hud_item_index
        cmp !hud_item_index_missile
        beq .missile_is_selected
        lda !unselected
        sta !select_mask
        bra .else
.missile_is_selected:
        lda !selected
        sta !select_mask
.else:
        ;; display custom digits on row 1
        ldx !digit1
        lda HUD_digits_tilemap_row1,X
        ora !select_mask
        sta !row1_missile_tile0
        ldx !digit2
        lda HUD_digits_tilemap_row1,X
        ora !select_mask
        sta !row1_missile_tile1
        ldx !digit3
        lda HUD_digits_tilemap_row1,X
        ora !select_mask
        sta !row1_missile_tile2
        ;; display custom missile icon on row 2
        lda #$0049
        ora !select_mask
        sta !row2_missile_tile0
        lda #$004A
        ora !select_mask
        sta !row2_missile_tile1
        lda #$004B
        ora !select_mask
        sta !row2_missile_tile2
        ;; display vanilla digits on row 3
        lda !samus_missiles
        ldx !row3_missile_index
	jsr !draw_three_HUD_digits
       rts
 
draw_super:
	jsr extract_three_digits
	lda !hud_item_index
	cmp !hud_item_index_super
	beq .super_is_selected
	lda !unselected
	sta !select_mask
	bra .else
.super_is_selected:
	lda !selected
	sta !select_mask
.else:
        ;; display custom digits on row 1
	ldx !digit1
	beq .two_digits
	lda HUD_digits_tilemap_row1,X
	ora !select_mask
	sta !row1_super_tile0
.two_digits:
	ldx !digit2
	lda HUD_digits_tilemap_row1,X
	ora !select_mask
	sta !row1_super_tile1
	ldx !digit3
	lda HUD_digits_tilemap_row1,X
	ora !select_mask
	sta !row1_super_tile2
        ;; display custom super icon on row 2
        lda #$0034
        ora !select_mask
        sta !row2_super_tile0
        lda #$0035
        ora !select_mask
        sta !row2_super_tile1
        ;; display vanilla digits on row 3
	lda !samus_super_missiles
        ldx !row3_super_index_vanilla
	jsr super_missile_count_drawing_row3
        rts
 
draw_power_bomb:
        jsr extract_three_digits
        lda !hud_item_index
        cmp !hud_item_index_power_bomb
        beq .power_bomb_is_selected
        lda !unselected
        sta !select_mask
        bra .else
.power_bomb_is_selected:
        lda !selected
        sta !select_mask
.else:
	;; display custom digits on row 1
        ldx !digit1
	beq .two_digits
        lda HUD_digits_tilemap_row1,X
        ora !select_mask
        sta !row1_power_bomb_tile0
.two_digits:
        ldx !digit2
        lda HUD_digits_tilemap_row1,X
        ora !select_mask
        sta !row1_power_bomb_tile1
        ldx !digit3
        lda HUD_digits_tilemap_row1,X
        ora !select_mask
        sta !row1_power_bomb_tile2
        ;; display custom pb icon on row 2
        lda #$0036
        ora !select_mask
        sta !row2_power_bomb_tile0
        lda #$0037
        ora !select_mask
        sta !row2_power_bomb_tile1
        ;; display vanilla digits on row 3
	lda !samus_power_bombs
	ldx !row3_power_bomb_index_vanilla
	jsr power_bomb_count_drawing_row3
        rts

clear_power_bomb:
        lda #$2C0F
        sta !row1_power_bomb_tile0
        sta !row1_power_bomb_tile1
        sta !row1_power_bomb_tile2
        sta !row2_power_bomb_tile0 ; 7EC66A = 8308330
        sta !row2_power_bomb_tile1
        sta $7EC6AA
        sta $7EC6AC
        rts
clear_super:
        lda #$2C0F
        sta !row1_super_tile0
        sta !row1_super_tile1
        sta !row1_super_tile2
        sta !row2_super_tile0 ; 8308324
        sta !row2_super_tile1
        sta $7EC6A4
        sta $7EC6A6
        rts
clear_missile:
        lda #$2C0F
        sta !row1_missile_tile0
        sta !row1_missile_tile1
        sta !row1_missile_tile2
        sta !row2_missile_tile0
        sta !row2_missile_tile1
        sta !row2_missile_tile2


;; workings out
;!row1_missile_tile0      = $7EC61C = 8308252
;!row1_missile_tile1      = $7EC61E = 8308254 = 2
;!row1_missile_tile2      = $7EC620 = 8308256 = 2
;!row2_missile_tile0      = $7EC65C = 8308316 = 64 from 1st
;!row2_missile_tile1      = $7EC65E = 8308318
;!row2_missile_tile2      = $7EC660 = 8308320 
;
;!row1_missile_tile0      = $7EC61C
;!row1_missile_tile1      = $7EC61E
;!row1_missile_tile2      = $7EC620
; end workings out

        ;; Third row
        lda #$2C0F
        sta $7EC69C
        lda #$2C0F
        sta $7EC69E
        lda #$2C0F
        sta $7EC6A0
;        ldy #$99AF
;        jsr clear_tile
        rts

        
clear_tile:
;;; $9A4C: Clear 2x2  icon to HUD tilemap ;;;
{
   ;; Parameters:
   ;;     X: HUD tilemap index
   ;;     Y: Source address

   ; Called by:
   ;     $9A0E with X = 1Ch, Y = $99AF: Clear super missiles to HUD tilemap
   ;     $9A1E with X = 22h, Y = $99B7: Clear power bombs to HUD tilemap
   ;     $9A2E with X = 28h, Y = $99BF: Clear grapple to HUD tilemap
   ;     $9A3E with X = 2Eh, Y = $99C7: Clear x-ray to HUD tilemap
        LDA #$2C0F ;[$80:99AF]  ;\
        STA $C608,y ;[$7E:C624];| top
        STA $C60A,y ;[$7E:C626];/ top
        STA $C648,y ;[$7E:C664];| mid
        STA $C64A,y ;[$7E:C666];/ mid
        RTS
}

   
 
;;; Extract three digits
;;; Parameters:
;;;     A: Three digits number
;;; Return:
;;;     $12: first digit
;;;     $14: second digit
;;;     $16: third digit
extract_three_digits:
	sta !dividend ; a = (a / 100) * 2
	sep #$20      ; 8-bit mode
	lda #$64 ; 100
	sta !divisor
	pha 
	pla 
	pha 
	pla 
	rep #$20                ; 16-bit mode
	lda !quotient
	asl A ; x2
	sta !digit1
	lda !remainder
 
;;; Extract two digits
;;; Parameters:
;;;     A: Two digits number
;;; Return:
;;;     $14: first digit
;;;     $16: second digit
extract_two_digits:
	sta !dividend
	sep #$20
	lda #$0A
	sta !divisor
	pha 
	pla 
	pha 
	pla 
	rep #$20
	lda !quotient
	asl A
	sta !digit2
	lda !remainder
	asl A
	sta !digit3
	rts

;;; HUD digits tilemap for row 1
HUD_digits_tilemap_row1:
        ;;     0      1      2      3      4      5      6      7      8      9
	dw $0045, $003C, $003D, $003E, $003F, $0040, $0041, $0042, $0043, $0044

;;; next patch start address (msu1)
warnpc $80D02F

;;; 85/ MISSILE/SUPER/POWER BOMB message tilemaps
org $858851
	dw $280F, $280F, $280F
	
org $858891
	dw $3049, $304A, $304B
	
org $858951
	dw $280F, $280F, $280F

org $858993
	dw $3034, $3035

org $858A4F
        dw $280F, $280F

org $858A8F
        dw $3036, $3037

;;; 9A/$B200: Standard BG3 tiles ;;;
org $9AB542
        db $9F, $7F, $B2, $7F, $AE, $73, $AE, $73, $B2, $7F, $9F, $7F, $FF, $80, $FF, $01
        db $F1, $FE, $FD, $2E, $2B, $F6, $FB, $26, $2D, $FE, $F1, $FE, $FF, $01, $FF, $80
        db $83, $7F, $86, $7F, $8F, $7A, $8E, $7B, $87, $7E, $83, $7F, $FF, $80, $FF, $01
        db $C1, $FE, $61, $FE, $F1, $5E, $71, $DE, $E1, $7E, $C1, $FE, $FF, $01

org $9AB5C0
        db $3C, $C3, $7C, $9B, $7C, $BB, $7C, $9B, $3C, $DB, $3C, $DB, $3C, $C3, $00, $FF
        db $FE, $01, $FF, $7C, $7F, $86, $FF, $3C, $FF, $60, $FF, $7E, $FF, $00, $00, $FF
        db $FE, $01, $FF, $7C, $FF, $06, $7F, $BC, $FF, $06, $FF, $7C, $FE, $01, $00, $FF
        db $3E, $C1, $7E, $9D, $FE, $2D, $FF, $4C, $FF, $7E, $FF, $0C, $1E, $E1, $00, $FF
        db $FF, $00, $FF, $7C, $FE, $61, $FF, $7C, $FF, $06, $FF, $7C, $FE, $01, $00, $FF
        db $7E, $81, $FE, $3D, $FE, $61, $FF, $7C, $FF, $66, $FF, $3C, $7E, $81, $00, $FF
        db $FF, $00, $FF, $7E, $FF, $06, $3F, $CC, $3E, $D9, $3C, $DB, $3C, $C3, $00, $FF
        db $7E, $81, $FF, $3C, $FF, $66, $FF, $3C, $FF, $66, $FF, $3C, $7E, $81, $00, $FF
        db $7E, $81, $FF, $3C, $FF, $66, $FF, $3E, $7F, $86, $7F, $BC, $7E, $81, $00, $FF
        db $7E, $81, $FF, $3C, $FF, $66, $FF, $66, $FF, $66, $FF, $3C, $7E, $81, $00

org $9AB691
        db $80, $80, $7F, $87, $7F, $89, $7E, $89, $7E, $87, $7F, $80, $7F, $FF, $80, $FF
        db $00, $E3, $FF, $1F, $FE, $FF, $21, $E1, $3F, $1E, $FF, $E3, $FF, $FF, $00, $FF
        db $01, $F9, $FE, $F1, $1E, $21, $FE, $E1, $3E, $11, $FE, $F9, $FE, $FF, $01